Changed injection of container to session, added deprecated to container and added test

ahmetakbn opened the pull request:
<!-- THE PR TEMPLATE IS NOT AN OPTION. DO NOT DELETE IT! -->

<!--
    Specify which issues will be fixed/closed.
    Remove it if this is not related.
-->

Refs #1033

### Changelog

<!-- MANDATORY
    Fill the changelog part inside the code block.
    Remove unneeded sections.
    Follow this schema: http://keepachangelog.com/
-->

```markdown
### Added
- Added test for `Security/SessionDownloadStrategy`

### Deprecated
- `$container` property in `Security/SessionDownloadStrategy`

### Changed
- Injection of `Session` instead of the whole `Container` in `Security/SessionDownloadStrategy`
```

### Subject

See #1032 and #1033

OskarStark commented:
The BC way looks good @ahmetakbn 👍 

/cc @Soullivaneuh @greg0ire @core23

core23 commented on Security/SessionDownloadStrategy.php:
```diff
@@ -39,14 +47,25 @@ class SessionDownloadStrategy implements DownloadStrategyInterface
+     * @param int                        $times
      */
-    public function __construct(TranslatorInterface $translator, ContainerInterface $container, $times)
+    public function __construct(TranslatorInterface $translator, $container, $times)
```
Should we use `$containerOrSession` as method property?

OskarStark replied:
but if you extend it the signature is different...


but indeed, it is different now, too

greg0ire replied:
I'd even recommend using `$session`, to make the diff of the future PR smaller, and to make self-explanatory what is really the recommended way.

greg0ire replied:
> but if you extend it the signature is different...

@OskarStark Argument names do no matter at all : https://3v4l.org/MAgfG , because different names don't make the signatures incompatible (both object can still be used with the same arguments)

greg0ire replied:
Also , the `NEXT_MAJOR` marker should be used to indicate that type hinting needs to be added in the non-BC PR

core23 commented on Security/SessionDownloadStrategy.php:
```diff
@@ -14,12 +14,20 @@
     /**
+     * @var Session
+     */
+    protected $session;
```
Why `protected`? Should be `private`.

OskarStark replied:
Sounds good

greg0ire commented on Security/SessionDownloadStrategy.php:
```diff
@@ -39,14 +47,25 @@ class SessionDownloadStrategy implements DownloadStrategyInterface
+    public function __construct(TranslatorInterface $translator, $container, $times)
     {
+        if ($container instanceof ContainerInterface) {
+            @trigger_error(
```
To be marked with : 

```
NEXT_MAJOR : remove this block
```

greg0ire commented on Tests/Security/SessionDownloadStrategyTest.php:
```diff
@@ -0,0 +1,52 @@
+ * @author Ahmet Akbana <ahmetakbana@gmail.com>
+ */
+final class SessionDownloadStrategyTest extends \PHPUnit_Framework_TestCase
+{
```
Please add a test for the legacy way, may, and mark it with : 

```php
/**
 * @group legacy
 * NEXT_MAJOR: remove this method
 */
```

greg0ire replied:
public function testLegacyConstructorUsageStillWorks
{
        $translator = $this->getMock('Symfony\Component\Translation\TranslatorInterface');
        $container = $this->getMock('Something…\ContainerInterface');

        $container->expects($this->once())
            ->method('get')
            ->with('session')
            ->willReturn($this->getMock('Symfony\Component\HttpFoundation\Session\Session'));

        $strategy = new SessionDownloadStrategy($translator, $session, 0);
}

ahmetakbn commented:
the changes has been done and squashed.

greg0ire commented on Security/SessionDownloadStrategy.php:
```diff
@@ -14,12 +14,15 @@
     /**
      * @var ContainerInterface
+     *
+     * @deprecated Since version 3.x, will be removed in 4.0.
```
Please add the `NEXT_MAJOR` here too (basically, everywhere we must not forget to do something).

greg0ire replied:
ping @ahmetakbn

greg0ire replied:
Oh sorry didn't see below

greg0ire commented on Security/SessionDownloadStrategy.php:
```diff
@@ -39,14 +42,31 @@ class SessionDownloadStrategy implements DownloadStrategyInterface
+     * @param ContainerInterface|Session $container
+     * @param int                        $times
+     */
+    public function __construct(TranslatorInterface $translator, $container, $times)
```
You should try to make future diff of the non-BC PR minimal. Please rename `$container` here (and in the phpdoc too). As seen before, changing the variable name is not a BC-break

greg0ire commented on Security/SessionDownloadStrategy.php:
```diff
@@ -39,14 +42,31 @@ class SessionDownloadStrategy implements DownloadStrategyInterface
+                E_USER_DEPRECATED
+            );
+
+            $this->session = $container->get('session');
```
So here it will be `$session->get('session')`. The deprecated code will look weird…

greg0ire commented on Security/SessionDownloadStrategy.php:
```diff
@@ -39,14 +42,31 @@ class SessionDownloadStrategy implements DownloadStrategyInterface
+
+            $this->session = $container->get('session');
+        } else {
+            $this->session = $container;
```
…and here it will be `$this->session = $session`. The new, recommended code will not look weird.

ahmetakbn commented:
@greg0ire changed the injected param name to session, added NEXT_MAJOR information to property and updated the doc ( added also information how to add service in yaml format )

greg0ire commented:
LGTM

greg0ire commented:
We have a problem with Travis : it did not run any tests. We will ask you to rebase soon, when this is fixed.

greg0ire commented:
You may rebase now :)

mvhirsch commented on Resources/doc/reference/security.rst:
```diff
@@ -74,89 +74,97 @@ Let's create the following strategy : a media can be downloaded only once per se
+         * @param int                 $times
          */
-        public function __construct(TranslatorInterface $translator, ContainerInterface $container, $times)
+        public function __construct(TranslatorInterface $translator, Session $session, $times)
```
Instead of using the concrete implementation, please use the Interface `Symfony\Component\HttpFoundation\Session\SessionInterface` .

greg0ire replied:
Good catch @mvhirsch !

mvhirsch commented on Security/SessionDownloadStrategy.php:
```diff
@@ -39,14 +43,31 @@ class SessionDownloadStrategy implements DownloadStrategyInterface
+        if ($session instanceof ContainerInterface) {
+            @trigger_error(
+                'Using an instance of Symfony\Component\DependencyInjection\ContainerInterface is deprecated since 
+                version 3.x and will be removed in 4.0. Use Symfony\Component\HttpFoundation\Session\Session instead.',
```
As mentioned above, please use the SessionInterface here too.

mvhirsch commented on Security/SessionDownloadStrategy.php:
```diff
@@ -39,14 +43,31 @@ class SessionDownloadStrategy implements DownloadStrategyInterface
+
+            $this->session = $session->get('session');
+        } else {
+            $this->session = $session;
```
Shouldn't we check the type here? Since if it's not a ContainerInterface but instead a [insert random class name here], it will fail?

Something like: 

```
elseif ($session instanceof SessionInterface) {
    $this->session = $session
} else {
    throw new \InvalidArgumentException()
}
```

mvhirsch commented on Resources/doc/reference/security.rst:
```diff
@@ -74,89 +74,97 @@ Let's create the following strategy : a media can be downloaded only once per se
 .. code-block:: xml
 
-        <service id="sonata.media.security.session_strategy" class="Sonata\MediaBundle\Security\SessionDownloadStrategy" >
+        <service id="sonata.media.security.session_strategy" class="Sonata\MediaBundle\Security\SessionDownloadStrategy" scope="request">
```
Scopes will stop working on Symfony3, see https://symfony.com/blog/new-in-symfony-2-8-deprecating-scopes-and-introducing-shared-services

greg0ire replied:
Shall @ahmetakbn write this definition dynamically, in php then ?

mvhirsch replied:
@greg0ire I'm sorry, I'm unaware of Sonatas Symfony3 migration process at the moment :-(
Since it requires more work, I think this should be done all at once (in an other PR)?

greg0ire replied:
Not sure I undestand what you mean. What I'm saying is that we should detect support for shared services and use the right service definition depending on that. Building the service definition should be done in https://github.com/sonata-project/SonataMediaBundle/blob/3.x/DependencyInjection/SonataMediaExtension.php

mvhirsch replied:
Yes, that's correct. I'm just wondering: are there other services in SonataMediaBundle which depends on scopes? If so, we should rebuilt them all at once in a new PR after this one.

greg0ire replied:
Oh but this is only documentation not code, I did not notice until now! So no need for BC system here, just add another example (or change it) please @ahmetakbn

mvhirsch commented:
@ahmetakbn good job! :+1:

ahmetakbn commented:
Opps, I did rebase to wrong branch again!

ahmetakbn commented:
- In documentation I added an example session strategy which allows given users to download media files.
- Added also missing service for session download strategy

I also need to add translations for session strategy, but how should I add the translations?

greg0ire commented on Resources/doc/reference/security.rst:
```diff
@@ -64,99 +64,97 @@ The Strategy class must implement the ``DownloadStrategyInterface`` which contai
-        private function getSession()
-        {
-            return $this->container->get('session');
+            return $this->translator->trans('description.users_download_strategy', array('%users%' => '<code>'.implode('</code>, <code>', $this->users).'</code>'), 'SonataMediaBundle');
```
Please linebreak this

greg0ire commented:
> I also need to add translations for session strategy, but how should I add the translations?

I don't know, I think they can be generated, @core23 , can you help? I think I've seen you commit lots of translations once, but I don't know how you do that…

greg0ire commented:
ping @core23

core23 commented:
> I don't know, I think they can be generated, @core23 , can you help? I think I've seen you commit lots of translations once, but I don't know how you do that…

Haven't figured out any kind of automatation for the translation files. I created them step by step...

OskarStark commented:
> Haven't figured out any kind of automatation for the translation files. I created them step by step...

me too

OskarStark commented:
@ahmetakbn you need to add them manually

ahmetakbn commented:
> @ahmetakbn you need to add them manually

done.

OskarStark commented on Resources/translations/SonataMediaBundle.en.xliff:
```diff
@@ -78,6 +78,10 @@
             </trans-unit>
+            <trans-unit id="description.session_download_strategy">
+                <source>description.session_download_strategy</source>
+                <target>The media can be retrieved %times% time(s) per session</target>
```
please add a `.` (dot) at the end of the sentence

greg0ire replied:
Also, can you use [pluralization](http://symfony.com/doc/current/components/translation/usage.html#pluralization) ?

ahmetakbn replied:
Like this?

greg0ire commented on Resources/translations/SonataMediaBundle.en.xliff:
```diff
@@ -78,6 +78,10 @@
             </trans-unit>
+            <trans-unit id="description.session_download_strategy">
+                <source>description.session_download_strategy</source>
+                <target>The media can be retrieved one time per session.|The media can be retrieved %times% times per session.</target>
```
This is fine.

greg0ire commented on Resources/doc/reference/security.rst:
```diff
@@ -64,99 +64,101 @@ The Strategy class must implement the ``DownloadStrategyInterface`` which contai
-        private function getSession()
-        {
-            return $this->container->get('session');
+            return $this->translator->trans(
```
You have to use `transChoice` here

ahmetakbn replied:
This is the doc file for an example strategy. There is no translation for this. And here the translation sentence will be something like this:
```
The media can be retrieved by the following users : %users%.
```
Just like for the roles:
```
The media can be retrieved by users with the following roles : %roles%.
```
IMO, there is no need to add the `pluralization` here.

greg0ire replied:
Oh right, I thought you had forgotten to call `transChoice` for the other call because you seem confused about it. Nevermind this comment.

OskarStark commented on Security/SessionDownloadStrategy.php:
```diff
@@ -77,14 +97,11 @@ public function isGranted(MediaInterface $media, Request $request)
-    private function getSession()
-    {
-        return $this->container->get('session');
+        return $this->translator->transChoice(
```
👍

OskarStark commented on Security/SessionDownloadStrategy.php:
```diff
@@ -14,12 +14,16 @@
 use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
 use Symfony\Component\Translation\TranslatorInterface;
 
```
i would add an author tag here, too, because you reworked a lot of this file

what do you think @greg0ire

ahmetakbn replied:
added the author tag.

OskarStark commented:
I think were good to go!

@ahmetakbn can you please squash your commits into one?

@greg0ire can you please give us a final review?

greg0ire commented on Security/SessionDownloadStrategy.php:
```diff
@@ -39,14 +48,34 @@ class SessionDownloadStrategy implements DownloadStrategyInterface
+        } elseif ($session instanceof SessionInterface) {
+            $this->session = $session;
+        } else {
+            throw new \InvalidArgumentException();
```
Please provide a message : `$session should be an instance of either ContainerInterface or Session Interface`

OskarStark replied:
good catch!

greg0ire replied:
RTM after that.

OskarStark commented:
Thank you for squashing @ahmetakbn 

@greg0ire feel free to push the buttton! 😄

greg0ire commented:
Thanks a lot @ahmetakbn , good job!

greg0ire commented:
@ahmetakbn : I changed your message because it was too long, but next time try to make several commits (translation and injection are very different concerns)

OskarStark commented:
I told him to suqash

greg0ire commented:
> I told him to suqash

Oh right… why ?

OskarStark commented:
> Oh right… why ?

i thought there were to many commits, but i think @ahmetakbn was right with ~3 commits. Sorry, will check this twice the next time!

Soullivaneuh commented:
@ahmetakbn An upgrade note is missing for deprecated feature. Could you please open another PR for that?

ahmetakbn commented:
@Soullivaneuh How can I add an `upgrade note`? this is the first time I faced this.

ahmetakbn commented:
> @Soullivaneuh How can I add an upgrade note? this is the first time I faced this.

ping @Soullivaneuh

greg0ire commented:
You need to describe what people need to do to upgrade in this file : https://github.com/sonata-project/SonataMediaBundle/blob/3.x/UPGRADE-3.x.md
